File: DSA/Python/src/dsa/lib/math/nums/tests/test_three_sum.py
import unittest
from typing import List

from dsa.lib.math.nums.three_sum import ThreeSum


class ThreeSumTest(unittest.TestCase):
    def test_case1(self):
        self.assertEqual([[0, 0, 0]], self.three_sum([0, 0, 0, 0]))
        self.assertEqual([[-2, 0, 2]], self.three_sum([-2, 0, 0, 2, 2]))
        self.assertEqual([[-1, 0, 1]], self.three_sum([-1, 0, 1, 0]))

    def test_case2(self):
        self.assertIn([-1, -1, 2], [[-1, -1, 2], [-1, 0, 1]])
        self.assertIn([-1, 0, 1], [[-1, -1, 2], [-1, 0, 1]])

    @staticmethod
    def three_sum(nums: List[int]):
        return ThreeSum().threeSum(nums)

File: DSA/Cpp/test/dsa/lib/math/nums/test_three_sum.cpp
#include <vector>
#include <dsa/lib/math/nums/three_sum.hpp>
#include "nums_test.h"
namespace dsa::lib::math::nums {
namespace {

class ThreeSumTest : public NumsTest {
 protected:
    static std::vector<std::vector<int>> threeSum(std::vector<int> nums) {
        return ThreeSum::threeSum(nums);
    };
};
TEST_F(ThreeSumTest, case1) {
    std::vector<int> nums = {-1, 0, 1, 2, -1, 4};
    std::vector<std::vector<int>> expected = {{-1, -1, 2}, {-1, 0, 1}};
    std::vector<std::vector<int>> actual = threeSum(nums);
    EXPECT_EQ(expected, actual);
}
TEST_F(ThreeSumTest, case2) {
    std::vector<int> nums = {};
    std::vector<std::vector<int>> expected = {};
    std::vector<std::vector<int>> actual = threeSum(nums);
    EXPECT_EQ(expected, actual);
}
TEST_F(ThreeSumTest, case3) {
    std::vector<int> nums = {0};
    std::vector<std::vector<int>> expected = {};
    std::vector<std::vector<int>> actual = threeSum(nums);
    EXPECT_EQ(expected, actual);
}

}
}

